### PR TITLE
dialects: (stablehlo) add stablehlo.dot

### PR DESCRIPTION
**Pull request by superlopuh**
This is the attribute used to specify along which dimensions the dot products happen.

CC @efferifick

### CODE REVIEW

File: xdsl/dialects/stablehlo.py [hunk 1]
```diff
@@ -77,6 +91,88 @@ class PrecisionAttr(EnumAttribute[Precision], SpacedOpaqueSyntaxAttribute):
...
+    lhs_batching_dimensions: ParameterDef[ArrayAttr[IntegerAttr[I64]]]
+    rhs_batching_dimensions: ParameterDef[ArrayAttr[IntegerAttr[I64]]]
+    lhs_contracting_dimensions: ParameterDef[ArrayAttr[IntegerAttr[I64]]]
+    rhs_contracting_dimensions: ParameterDef[ArrayAttr[IntegerAttr[I64]]]
+
+    def print_parameters(self, printer: Printer) -> None:
+        with printer.in_angle_brackets():
+            with printer.indented():
+                printer.print_string("\nlhs_batching_dimensions = [")
+                printer.print_list(
```
Comment (alexarice) [inline xdsl/dialects/stablehlo.py:113]: Does directly printing the `ArrayAttr` not work here?

File: xdsl/dialects/stablehlo.py [hunk 2]
```diff
@@ -77,6 +91,88 @@ class PrecisionAttr(EnumAttribute[Precision], SpacedOpaqueSyntaxAttribute):
...
+                    lambda dim: printer.print_string(f"{dim.value.data}"),
+                )
+            printer.print_string("]\n")
+
+    @classmethod
+    def parse_parameters(cls, parser: AttrParser) -> Sequence[Attribute]:
+        with parser.in_angle_brackets():
+            parser.parse_characters("lhs_batching_dimensions")
+            parser.parse_punctuation("=")
+            lhs_batching_dimensions = parser.parse_comma_separated_list(
```
Comment (alexarice) [inline xdsl/dialects/stablehlo.py:142]: This would probably ideally be packaged up into a helper function

**Review by alexarice — COMMENTED** [review]
Small stylistic nits.

File: xdsl/dialects/stablehlo.py (hunk 1 above)
Comment (superlopuh) [reply xdsl/dialects/stablehlo.py:113]: It wants to print the types for each int, I'll look into a more elegant way of doing this

File: xdsl/dialects/stablehlo.py [hunk 3]
```diff
@@ -105,6 +112,93 @@ class TokenType(TypeAttribute, ParametrizedAttribute):
...
+    rhs_contracting_dimensions: ParameterDef[ArrayAttr[IntegerAttr[I64]]]
+
+    @staticmethod
+    def _print_parameter(
+        name: str, value: ArrayAttr[IntegerAttr[I64]], printer: Printer
+    ):
+        printer.print_string(f"\n{name} = [")
+        printer.print_list(
+            value.data,
+            lambda dim: printer.print_string(f"{dim.value.data}"),
```
Comment (alexarice) [inline xdsl/dialects/stablehlo.py:137]:
```suggestion
            lambda dim: printer.print(dim.value.data),
```

**Review by alexarice — APPROVED** [review]

File: xdsl/dialects/stablehlo.py (hunk 3 above)
Comment (alexarice) [reply xdsl/dialects/stablehlo.py:137]: Can ignore this if you want

Comment (superlopuh) [reply xdsl/dialects/stablehlo.py:137]: I actually prefer print_string to print, which is unfortunately pretty slow right now

**Pull request merged**